drivers/wdm/audio/backpln/portcls/pin_wavepci.cpp

```diff
@@ -588,22 +646,33 @@ CPortPinWavePci::DeviceIoControl(
 
     IoStack = IoGetCurrentIrpStackLocation(Irp);
 
-    if (IoStack->Parameters.DeviceIoControl.IoControlCode == IOCTL_KS_PROPERTY)
+    switch (IoStack->Parameters.DeviceIoControl.IoControlCode)
     {
-        return HandleKsProperty(Irp);
-    }
-    else if (IoStack->Parameters.DeviceIoControl.IoControlCode == IOCTL_KS_WRITE_STREAM || IoStack->Parameters.DeviceIoControl.IoControlCode == IOCTL_KS_READ_STREAM)
-    {
-       return HandleKsStream(Irp);
-    }
+        case IOCTL_KS_PROPERTY:
+            return HandleKsProperty(Irp);
 
-    UNIMPLEMENTED;
+        case IOCTL_KS_ENABLE_EVENT:
+            return PcHandleEnableEventWithTable(Irp, &m_Descriptor);
+
+        case IOCTL_KS_DISABLE_EVENT:
+            return PcHandleDisableEventWithTable(Irp, &m_Descriptor);
+
+        case IOCTL_KS_RESET_STATE:
+            return HandleKsReset(Irp);
+
+        case IOCTL_KS_READ_STREAM:
+        case IOCTL_KS_WRITE_STREAM:
+            return HandleKsStream(Irp);
+
+        default:
+            return KsDefaultDeviceIoCompletion(DeviceObject, Irp);
+    }
 
     Irp->IoStatus.Information = 0;
```

Comment:
The control never reaches here, you have a `return` statement in every case of `switch`

Reply:
Indeed. So I should use an `NTSTATUS` variable.
I tried to use it from the start, but it caused a bugcheck 0x44 (MULTIPLE_IRP_COMPLETE_REQUESTS), so then it probably needs to remove other IRP completions from all sub-functions called from `DeviceIoControl`.

Reply:
You can just move the `default` case out of the switch. Alternatively you can probably remove this code and the compiler shouldn't complain because it's smart enough to know that every switch case returns.